app/src/main/java/de/westnordost/streetcomplete/quests/building_entrance_reference/AddEntranceReference.kt

```diff
@@ -0,0 +1,101 @@
+import de.westnordost.streetcomplete.quests.building_entrance_reference.FlatRange
+import de.westnordost.streetcomplete.quests.building_entrance_reference.ReferenceCode
+import de.westnordost.streetcomplete.quests.building_entrance_reference.ReferenceCodeAndFlatRange
+import de.westnordost.streetcomplete.quests.building_entrance_reference.Unsigned
+
+class AddEntranceReference : OsmElementQuestType<EntranceAnswer> {
+
+    private val buildingFilter by lazy { """
+        ways with
+          building = apartments
+          and ruins != yes and disused != yes
+    """.toElementFilterExpression() }
+
+    override val changesetComment = "Add entrance identification"
+    override val wikiLink = "Key:ref"
+    override val icon = R.drawable.ic_quest_door_address
+    override val questTypeAchievements = listOf(CITIZEN, BLIND)
+    override val enabledInCountries = NoCountriesExcept("BG", "PL", "RU", "UA", "BY", "EE", "RO")
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_entrance_reference
+
+    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
+        getMapData().filter { it.tags.containsKey("entrance") }.asSequence()
+
+    override fun isApplicableTo(element: Element): Boolean? =
+        if (element !is Node || (element.tags["entrance"] != "main" && element.tags["entrance"] != "staircase") ) false else null
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        // note: it does not support multipolygon buildings
+        val buildingWays = mapData.ways.asSequence()
+            .filter { buildingFilter.matches(it) }
+        val returned = mutableListOf<Node>()
+
+        for (way in buildingWays) {
+            val entrancesForApartments = mutableListOf<Node>()
+            way.nodeIds.forEach { nodeId ->
+                val node = mapData.getNode(nodeId)
+                if(node != null) {
+                    if(node.tags["entrance"] == "main" || node.tags["entrance"] == "staircase") {
+                        entrancesForApartments.add(node)
+                    }
+                }
+            }
+            if(entrancesForApartments.count() >= 2) {
+                for (entrance in entrancesForApartments) {
+                    if(entrance.tags["noref"] == "yes") {
+                        continue
+                    }
+                    if(entrance.tags["ref:signed"] == "no") {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("addr:flats")) {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("name")) {
+                        continue
+                    }
+                    if(entrance.tags.containsKey("ref")) {
+                        continue
+                    }
+                    returned.add(entrance)
+                }
+            }
+        }
+        return returned
+    }
```

Comment:
```suggestion
    // TODO these vals should be at the top of the file actually
    private val entrancesFilter by lazy { 
        "nodes with entrance ~ main|staircase".toElementFilterExpression()
    }

    private val noEntranceRefFilter by lazy { """
        nodes with
          and noref != yes
          and ref:signed != no
          and !addr:flats
          and !name
          and !ref
    """.toElementFilterExpression() }

    override fun isApplicableTo(element: Element): Boolean? =
        if (!entrancesFilter.matches(element)) false else null

    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
        // note: it does not support multipolygon buildings
        val buildings = mapData.ways.asSequence()
            .filter { buildingFilter.matches(it) }

        val result = mutableListOf<Node>()
        for (building in buildings) {
            val buildingEntrances = building.nodeIds
                .mapNotNull { mapData.getNode(it) }
                .filter { entrancesFilter.matches(it) }

            if (buildingEntrances.count() < 2) continue

            result.addAll(buildingEntrances.filter { noEntranceRefFilter.matches(entrance) })
        }
        return result
    }
```

I understand that `flatMap` may not be easier to understand when one is not thinking in functional. I suggest this, because:
- the filter syntax is easily readable, better readable then code that does the same
- using the filter syntax for such things is consistent with the rest of the code base
- naming: usually we use `result`, not `returned`
- it's also shorter
- less deep nesting IMO improves readability